packages/core/crates/core-network/src/network.rs

```diff
@@ -0,0 +1,905 @@
+                } else {
+                    true
+                }
+            } else {
+                false
+            };
+
+            if ! has_entry && ! is_ignored {
+                let mut entry = PeerStatus::new(peer.clone(), origin);
+                entry.is_public = self.network_actions_api.is_public(&peer);
+                self.refresh_network_status(&entry);
+
+                if let Some(_x) = self.entries.insert(peer.to_string(), entry) {
+                    // warn!("Evicting an existing record for {}, this should not happen!", &x);
+                }
+            }
+        }
+    }
+
+    /// Remove PeerId from the network
+    pub fn remove(&mut self, peer: &PeerId) {
+        self.prune_from_network_status(&peer);
+        self.entries.remove(peer.to_string().as_str());
+        // TODO: remove from ignored and excluded as well?
+    }
+
+    /// Update the PeerId record in the network
+    pub fn update(&mut self, peer: &PeerId, ping_result: crate::types::Result) {
+        if let Some(existing) = self.entries.get(peer.to_string().as_str()) {
+            let mut entry = existing.clone();
+            entry.last_seen = if ping_result.is_err() {current_timestamp()} else {ping_result.ok().unwrap()} ;
+            entry.heartbeats_sent = entry.heartbeats_sent + 1;
+            // entry.is_public = self.network_actions_api.is_public(&peer);    // TODO: Reconsider whether this is necessary
+
+            if ping_result.is_err() {
+                entry.backoff = MAX_BACKOFF.max(entry.backoff.powf(BACKOFF_EXPONENT));
+                entry.quality = 0.0_f64.max(entry.quality - QUALITY_STEP);
+
+                if entry.quality < (QUALITY_STEP / 2.0) {
+                    self.network_actions_api.close_connection(&entry.id);
+                    self.prune_from_network_status(&entry.id);
+                    self.entries.remove(entry.id.to_string().as_str());
+                    return
+                }
+
+                if entry.quality < BAD_QUALITY {
+                    self.ignored.insert(entry.id.to_string(), current_timestamp());
+                    // self.entries.remove(entry.id.to_string().as_str());
+                    // self.prune_from_network_status(&entry.id);
+                    // TODO: Just add the entry to ignored? Prune once 0.0 quality is reached?
+                    return
+                }
+
+                if entry.quality < self.network_quality_threshold {
+                    self.network_actions_api.on_peer_offline(&entry.id);
+                }
+            } else {
+                entry.heartbeats_succeeded = entry.heartbeats_succeeded + 1;
+                entry.backoff = MIN_BACKOFF;
+                entry.quality = 1.0_f64.min(entry.quality + 0.1)
+            }
+
+            self.refresh_network_status(&entry);
+            self.entries.insert(entry.id.to_string(), entry);
+        } else {
+            info!("Ignoring update request for unknown peer {:?}", peer);
+        }
```

Comment:
@robertkiel @tolbrino @NumberFour8 
This could be done better, the logic is strange, how could it be improved? Drop connection on `BAD_QUALITY`?

Reply:
Also, during the runtime the ingored list just gets longer and longer, with no purgatory mechanism, so eventually, given enought runtime... the node could get out of friends to talk to, as they will all be ignored, because they might have been unavailable.

This is also true, if the current node, that performs the pings, suffers from a network issue, it will put everybody it cannot reach into an ignore list and never contact them again. This logic is not sustainable on long running daemons.

Reply:
@robertkiel I'd strongly suggest we revisit this entire strategy.